File: smash-ranks-android/app/src/main/java/com/garpr/android/models/RankingsBundle.java
package com.garpr.android.models;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;

public class RankingsBundle implements Parcelable {

    @Nullable
    @SerializedName("ranking")
    private ArrayList<Ranking> mRankings;

    @Nullable
    @SerializedName("tournaments")
    private ArrayList<String> mTournaments;

    @SerializedName("time")
    private SimpleDate mTime;

    @SerializedName("id")
    private String mId;

    @SerializedName("region")
    private String mRegion;


    @Override
    public boolean equals(final Object obj) {
        return obj instanceof RankingsBundle && mId.equals(((RankingsBundle) obj).getId());
    }

    public String getId() {
        return mId;
    }

    @Nullable
    public ArrayList<Ranking> getRankings() {
        return mRankings;
    }

    public SimpleDate getTime() {
        return mTime;
    }

    @Nullable
    public ArrayList<String> getTournaments() {
        return mTournaments;
    }

    @Override
    public int hashCode() {
        return mId.hashCode();
    }

    public boolean hasPreviousRank() {
        if (hasRankings()) {
            // noinspection ConstantConditions
            for (final Ranking ranking : mRankings) {
                if (ranking.getPreviousRank() != null) {
                    return true;
                }
            }
        }

        return false;
    }

    public boolean hasRankings() {
        return mRankings != null && !mRankings.isEmpty();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(final Parcel dest, final int flags) {
        dest.writeTypedList(mRankings);
        dest.writeParcelable(mTime, flags);
        dest.writeString(mId);
        dest.writeString(mRegion);
        dest.writeStringList(mTournaments);
    }

    public static final Creator<RankingsBundle> CREATOR = new Creator<RankingsBundle>() {
        @Override
        public RankingsBundle createFromParcel(final Parcel source) {
            final RankingsBundle rb = new RankingsBundle();
            rb.mRankings = source.createTypedArrayList(Ranking.CREATOR);
            rb.mTime = source.readParcelable(SimpleDate.class.getClassLoader());
            rb.mId = source.readString();
            rb.mRegion = source.readString();
            rb.mTournaments = source.createStringArrayList();
            return rb;
        }

        @Override
        public RankingsBundle[] newArray(final int size) {
            return new RankingsBundle[size];
        }
    };

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/models/RankingsBundle.kt
package com.garpr.android.models

import android.os.Parcel
import android.os.Parcelable
import com.garpr.android.extensions.createParcel
import com.garpr.android.extensions.readAbsRegion
import com.garpr.android.extensions.requireParcelable
import com.garpr.android.extensions.requireString
import com.garpr.android.extensions.writeAbsRegion
import com.google.gson.annotations.SerializedName

data class RankingsBundle(
        @SerializedName("ranking") val rankings: List<RankedPlayer>? = null,
        @SerializedName("tournaments") val tournaments: List<String>? = null,
        @SerializedName("ranking_criteria") val rankingCriteria: AbsRegion,
        @SerializedName("time") val time: SimpleDate,
        @SerializedName("id") val id: String,
        @SerializedName("region") val region: String
) : Parcelable {

    companion object {
        @JvmField
        val CREATOR = createParcel {
            RankingsBundle(
                    it.createTypedArrayList(RankedPlayer.CREATOR),
                    it.createStringArrayList(),
                    it.readAbsRegion(),
                    it.requireParcelable(SimpleDate::class.java.classLoader),
                    it.requireString(),
                    it.requireString()
            )
        }
    }

    override fun equals(other: Any?): Boolean {
        return other is RankingsBundle && id.equals(other.id, ignoreCase = true)
    }

    override fun hashCode() = id.hashCode()

    fun hasPreviousRank(): Boolean {
        return rankings?.asSequence()
                ?.firstOrNull { it.previousRank != null } != null
    }

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeTypedList(rankings)
        dest.writeStringList(tournaments)
        dest.writeAbsRegion(rankingCriteria, flags)
        dest.writeParcelable(time, flags)
        dest.writeString(id)
        dest.writeString(region)
    }

}
